test/gtest/dropout_ocl_hip.cpp

```diff
@@ -0,0 +1,825 @@
+    size_t total_work = (in_len[4] / RD_BLCK) * in_len[3] * in_len[2] * in_len[1] * in_len[0];
+
+    size_t max_wk_grp = use_prng ? std::min(size_t(MAX_PRNG_STATE), handle.GetImage3dMaxWidth())
+                                 : size_t(MAX_WORKITEM_NUM);
+    size_t wk_grp_num =
+        std::min(max_wk_grp / 256,
+                 ((in_len[4] * in_len[3] * in_len[2] * in_len[1] * in_len[0] + 255) / 256));
+
+    if(use_prng)
+    {
+        size_t states_num = stateSizeInBytes / sizeof(rocrand_state_xorwow);
+        if(states_num < wk_grp_num * 256)
+        {
+            MIOPEN_THROW("Insufficient state size for parallel PRNG");
+        }
+    }
+
+    std::string network_config =
+        "bwd-" + std::string(dyDesc.GetType() == miopenHalf ? "fp16-" : "fp32-") + "-seed" +
+        std::to_string(seed) + "-rng" + std::to_string(rng_mode) + "-prng" +
+        std::to_string(static_cast<int>(use_prng)) + "-evo" +
+        std::to_string(static_cast<int>(state_evo)) + "-blk" + std::to_string(RD_BLCK) + "-wg" +
+        std::to_string(wk_grp_num) /* + "-noise" + std::to_string(noise_shape.GetLengths()[0]) */;
+
+    std::string program_name;
+    std::string kernel_name;
+
+    if(!use_hip)
+    {
+
+        program_name = "MIOpenDropout.cl";
+        kernel_name  = "DropoutBackward";
+        network_config += "-ocl";
+    }
+    else
+    {
+
+        program_name = "MIOpenDropoutHIP.cpp";
+        kernel_name  = "DropoutBW";
+        network_config += "-hip";
+    }
+
+    // TODO: Add noise shape
+    // for(int i = 1; i < noise_shape.GetNumDims(); i++)
+    //    network_config += "x" + std::to_string(noise_shape.GetLengths()[i]);
+
+    float amp_scale = miopen::float_equal(dropout, 1.0) ? 0 : 1 / (1 - dropout);
+
+    std::string params;
+
+    const std::string data_type = miopen::GetDataType(dyDesc.GetType());
+    const std::string READ_DAT_TYPE =
+        RD_BLCK == 1 ? data_type : data_type + std::to_string(RD_BLCK);
+
+    params += " -DRD_BLCK=" + std::to_string(RD_BLCK) + " -DREAD_DAT_TYPE=" + READ_DAT_TYPE +
+              " -DREAD_BOOL_TYPE=" +
+              std::string(RD_BLCK == 4   ? "uint"
+                          : RD_BLCK == 2 ? "ushort"
+                                         : "uchar");
+
+    if(use_prng)
+    {
+        params += " -DUSE_PRNG=1";
+    }
+
+    if(dyDesc.GetType() == miopenHalf)
+        params += " -DMIOPEN_USE_FP16=1";
+    else
+        params += " -DMIOPEN_USE_FP32=1";
+
+    params += " -DRUN_FORWARD=0";
+
+    const std::vector<size_t> vld{256, 1, 1};
+    const std::vector<size_t> vgd{wk_grp_num * 256, 1, 1};
+
+    handle.AddKernel(kernel_name, network_config, program_name, kernel_name, vld, vgd, params)(
+        pstates,
+        dropout,
+        amp_scale,
+        static_cast<int>(in_len[1]),
+        static_cast<int>(in_len[2]),
+        static_cast<int>(in_len[3]),
+        static_cast<int>(in_len[4]),
+        dy,
+        static_cast<int>(out_str[0]),
```

Comment:
take strides as `size_t` because we might need to support `large tensors` > 2GB.

Reply:
It seems `size_t` based indexing introduces significant perf degradation.
The way how to do it properly has to be evaluated and thoroughly tested. I guess it has to be a separated task.

Reply:
But what is needed to do right away is narrowing the applicability, correct?

Reply:
> But what is needed to do right away is narrowing the applicability, correct?

Sure, but we don't need `size_t` for small tensors. Surely we can use 64bit indexing for the big ones, but it's a separated task.